Sources/TuistAutomation/Device/DeviceController.swift

```diff
@@ -0,0 +1,168 @@
+    func findAvailableDevices() async throws -> [PhysicalDevice]
+    func installApp(
+        at path: AbsolutePath,
+        device: PhysicalDevice
+    ) async throws
+    func launchApp(
+        bundleId: String,
+        device: PhysicalDevice
+    ) async throws
+}
+
+public final class DeviceController: DeviceControlling {
+    private let fileSystem: FileSysteming
+    private let commandRunner: CommandRunning
+
+    public init(
+        fileSystem: FileSysteming = FileSystem(),
+        commandRunner: CommandRunning = CommandRunner()
+    ) {
+        self.fileSystem = fileSystem
+        self.commandRunner = commandRunner
+    }
+
+    public func findAvailableDevices() async throws -> [PhysicalDevice] {
+        try await fileSystem.runInTemporaryDirectory(prefix: "device-controller-find-available-devices") { temporaryPath in
+            let devicesListOutputPath = temporaryPath.appending(component: "devices-list-output-path.json")
+            _ = try await commandRunner.run(
+                arguments: [
+                    "/usr/bin/xcrun", "devicectl",
+                    "list", "devices",
+                    "--json-output", devicesListOutputPath.pathString,
+                ]
+            )
+            .concatenatedString()
+
+            let deviceList = try JSONDecoder().decode(
+                DeviceList.self,
+                from: try await fileSystem.readFile(at: devicesListOutputPath)
+            )
+
+            return deviceList.result.devices
+                .map(PhysicalDevice.init)
+        }
+    }
+
+    public func installApp(
+        at path: AbsolutePath,
+        device: PhysicalDevice
+    ) async throws {
+        logger.debug("Installing app at \(path) on simulator device with id \(device.id)")
+        do {
+            _ = try await commandRunner.run(
+                arguments: [
+                    "/usr/bin/xcrun", "devicectl",
+                    "device", "install", "app",
+                    "--device", device.id,
+                    path.pathString,
+                ]
+            )
+            .concatenatedString()
+        } catch let error as CommandError {
+            if case let .terminated(_, stderr) = error, stderr.contains("ApplicationVerificationFailed") {
+                throw DeviceControllerError.applicationVerificationFailed
+            } else {
+                throw error
+            }
+        }
+    }
+
+    public func launchApp(
+        bundleId: String,
+        device: PhysicalDevice
+    ) async throws {
+        logger.debug("Launching app with bundle id \(bundleId) on a physical device with id \(device.id)")
+        _ = try await commandRunner.run(
+            arguments: [
+                "/usr/bin/xcrun", "devicectl",
+                "device", "process", "launch",
+                "--device", device.id,
+                bundleId,
+            ]
+        )
+        .concatenatedString()
+    }
+}
+
+private struct DeviceList: Codable {
```

Comment:
What about flattening this such that?

```swift
typelias DeviceList = [Device]
```

Reply:
I can't really do that since `DeviceList` doesn't contain the `devices` array directly but the array is in the `results` property. It's a private struct used only for decoding the json from `devicectl`, so I'd say let's keep it as-is.